### Data Consistency Checks

## 01. Importing Libraries

In [5]:
#import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data

In [6]:
df_prods = pd.read_csv(r'/Users/melaniehouston/Library/CloudStorage/OneDrive-Personal/Career Foundry/ 04 2024 Instacart Basket Analysis/02 Data/Original Data/products.csv', index_col = False)

In [7]:
df_ords = pd.read_csv(r'/Users/melaniehouston/Library/CloudStorage/OneDrive-Personal/Career Foundry/ 04 2024 Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

In [8]:
path = r'/Users/melaniehouston/Library/CloudStorage/OneDrive-Personal/Career Foundry/ 04 2024 Instacart Basket Analysis'

### 03. Mixed-Type Data

In [9]:
#create a dataframe
df_test = pd.DataFrame()

In [10]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
#Check for mixed type columns
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [13]:
# converting mixed columns to 'str'
df_test['mix'] = df_test['mix'].astype('str')

In [14]:
df_test.dtypes #proving mix dtype is object or string

mix    object
dtype: object

### 04. Missing Values

In [15]:
#showing any missing values for the products dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
#creating a subset to list rows where the product_name is null 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
#this will show rows & columns
df_prods.shape

(49693, 5)

In [19]:
#Creating a subset of df_prods_clean where null product names won't appear
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [20]:
#this will show rows & columns, with 16 fewer rows
df_prods_clean.shape

(49677, 5)

### 05. Duplicates

In [21]:
# making new subset containing ONLY rows that are duplicated in the df_prods_clean subset
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
#checking rows and columns
df_prods_clean.shape

(49677, 5)

In [24]:
#creating df that DOESN'T include the duplicates we identified using drop_duplicates() function
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
#showing rows and columns with 5 fewer rows!
df_prods_clean_no_dups.shape

(49672, 5)

### 06 Export changes

In [26]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

### TASK 4.5 Data Consistency Checks

Step 2 Descriptive Statistics on df_ords

In [27]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  orders_time_of_day  days_since_prior_order  
count        3.421083e+06        3.421083e+06            3.214874e+06  
mean         2.776219e+00        1.345202e+01            1.111484e+01  
std          2.046829e+00        4.226088e+00            9.206737e+00  
min          0.000000e+00        0.000000e+00            0.000000e+00  
25%          1.000000e+00        1.000000e+01            4.000000e+00  
50%          3.000000e+00        1.300000e+01            7.000000e+00  
75%          5.000000e+00        1.600000e+01            1.500000e+01  
max          6.000000e+00        2.300000e+01            3.000000e+01

The minimum row only has answers of either zero or 1, for the time of day or day of the week, 'days since prior order'
is also 0, so this could mean same day? 
In the maximum row, the 'days of the week' is 6, days of week being 0-6, which is 7 days
I would like these to be displayed as regular numbers not scientific with the e+00, etc.

Step 3 Checking for mixed-type data

In [28]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

#Nothing returned here (?)

#Step 5 Running a check for missing values

In [29]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
orders_time_of_day             0
days_since_prior_order    206209
dtype: int64

#Days since prior order has 206209 values missing!

Step 6 Addressing missing values

In [30]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [31]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  orders_time_of_day  days_since_prior_order  
0                         2                   8                     NaN  
11                        2                  11                     NaN  
26                        1                  14                     NaN  
39                        6                  11                     NaN  
45                        3                  12                     NaN  
...                     ...                 ...                     ...  
3420930                   4                  12                     NaN  
3420934                   3                  18                     NaN  
3421002                   6                  19                     NaN  
3421019                   1                  15                     NaN  
3421069                   3                  11                     NaN  

[206209 rows x 8 columns]

In [32]:
#finding the average value of the days since prior order column
mean_value = df_ords['days_since_prior_order'].mean()

In [33]:
mean_value

11.114836226863012

In [34]:
df_ords['days_since_prior_order'].fillna(mean_value, inplace = True)

In [35]:
df_ords = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [36]:
df_ords[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  orders_time_of_day  days_since_prior_order  
0                         2                   8               11.114836  
11                        2                  11               11.114836  
26                        1                  14               11.114836  
39                        6                  11               11.114836  
45                        3                  12               11.114836  
...                     ...                 ...                     ...  
3420930                   4                  12               11.114836  
3420934                   3                  18               11.114836  
3421002                   6                  19               11.114836  
3421019                   1                  15               11.114836  
3421069                   3                  11               11.114836  

[206209 rows x 8 columns]

I calculated the mean of the days since prior order column and then used it with fillna to fill in the missing data in that column.

Step 7 checking for duplicates in df_ords

In [38]:
df_dups = df_ords[df_ords.duplicated()]

In [39]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, orders_time_of_day, days_since_prior_order]
Index: []

In [40]:
#There are no duplicates.  df.drop_duplicates() would remove them if they were present. 

#07 Exporting data

In [41]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))